README
                    _    _     ___     _  _    
        __ _  _  _ | |_ | |_  |_  ) __| || |__ 
       / _` || || ||  _|| ' \  / / / _` || '_ \
       \__,_| \_,_| \__||_||_|/___|\__,_||_.__/

-------------------------------------------------------------
Auth2DB (version 0.3.0)                           10/06/2010
-------------------------------------------------------------

Auth2DB is an Open Source application that stores records (logs) in a 
centralized way to facilitate a real time processing and usage, with 
an optimal performance. Also is capable to create reports, statistics 
and alerts.

This application has the advantage of a scalable design, consequently 
it is able to process any type of record and to manage their content 
from a Web panel. 

Auth2db allows you to centralize all server, clients and applications 
records in a single database, making record analysis more accessible, 
and alerting in case of finding a potential problem.


More details:
-------------

Test Distributions at the moment: 
Debian, Ubuntu, Slackware, Backtrack, Fedora, RedHat, openSUSE


Installation 0.3.x.tar.gz:
-------------------------

# install mysql-server-5.0
# install python-mysqldb

# install auth2db
 tar zxvf auth2db.0.3."version".tar.gz
 cd auth2db
 python setup.py install

# Create database "auth2db"
 mysql> create database auth2db;

# To create tables, config server and hosts exec as ROOT:
 auth2db-config

# EDIT /etc/auth2db/filters.conf to enable|disabled types or create other.

# Verify /etc/auth2db/auth2db.conf to enable|disabled alerts 
# (if you have many hosts, only 1 host alert activation is required to execute all).

# Use Daemon
# /etc/init.d/auth2db-daemon start|stop

# READY...

# execute auth2db or start daemon

http://localhost/auth2db 
(user:admin | pass:admin)


;)


Experimental UPDATE_IP_SSHD:
---------------------------
Actualiza las IPs de los registros "session opened", que se generan 
luego de ingresar por SSH, seguido del registro "Accepted".
Esta feature es experimental. Si nota que la performance cae,
desactive esta opcion en el archivo "/etc/auth2db/auth2db.conf".
Es necesario tener activo el filtro [sshd_session] para su funcionamiento.

# enabled
UPDATE_IP_SSHD = y

# Disabled
UPDATE_IP_SSHD = n



Configuration files:
--------------------

# PATH donde esta la configuracion principal
"/etc/auth2db/"

# PATH de modulos adicionales como configobj.py
"/usr/share/auth2db/modules"

# PATH de scripts adicionales como alert.py, cleaner.py, functions.py, etc
"/usr/share/auth2db/scripts"


Contact:
--------

You need some new features, you found bugs or you just
appreciate this program, you can contact me at :

Website: http://www.auth2db.com.ar

Ezequiel Vera: <ezequielvera [_at_] yahoo.com.ar>


License:
--------

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA


CHANGELOG
                    _    _     ___     _  _
        __ _  _  _ | |_ | |_  |_  ) __| || |__
       / _` || || ||  _|| ' \  / / / _` || '_ \
       \__,_| \_,_| \__||_||_|/___|\__,_||_.__/

-------------------------------------------------------------
Auth2DB (version 0.3.0)                           10/06/2010
-------------------------------------------------------------

ver. 0.3.0 (10/06/2010) - beta
 ----------
  + Search feature has been added.
  * General status has been modified.
  + Select Log Source feature has been added.
  * All the logs from source are saved now.
  * VIEW module has been changed.
  * Huge logs bug has been fixed.
  * ALERTS settings has been changed.
  * Floatting box in views has been fixed.
  + CORE error logs has been added.
  + Views settings has been added.
  + Thread checker has been added.
  + Regex creator and tester has been added.
  + Javascript colour picker has been added.
  + Modules with scripts has been added.

ver. 0.2.8 (22/04/2009) - beta
 ----------
  + Table tipo_action_config has been added.
  + Table list_config has been added.
  + Table list_item has been added.
  + Statistinc Graph has been updated.
  + New lists features has been added.
  + New features to alerts has been added.
  + New Security checks has been added.

ver. 0.2.7 (18/03/2009) - beta
 ----------
  + Threads to auth2db has been added.
  + Logs Tables per day.
  + New Security Checks.
  + Reports Alert Multiple Emails.
  + Role Access Control has been added.
  + New Chars library.
  - GD Library has been removed.

ver. 0.2.6 (14/10/2008) - beta
 ----------
  + Add Threads to auth2db. 
  + Threads -d param added.
  * Change Class Flag for Functions Flag (auth2db).
  - FLAG Files Deleted.
  - TMP Logs Files Deleted.

ver. 0.2.5 (30/07/2008) - beta
 ----------
  + Add SQL Injection and XSS control.
  + Add Realtime Row Limit.
  + Add New Filters.
  * Minor Fix in Advance View, Increase Performance.
  * Fix updateipsshd function, Increase Performance.
  - Disabled updateipsmbd function, Decrease Performance.

ver. 0.2.4 (20/04/2008) - beta
 ----------
  + Add Reports table.
  + Add Reports creator in frontend.
  + Add Reports View.
  + Update Settings Control Panel frontend (with report creator).
  + Report Graph renamed to Statistic in Menu.
  + Update frontend icons.
  * Fix Filters Remote Syslog Hostname

ver. 0.2.3 (17/04/2008) - beta
 ----------
  + Change frontend db connection.
  + Update filters, and flag.d path. 
  * FIX hardcoding year in frontend.
  * FIX hardcoding year in statistic graph.
	 
ver. 0.2.2 (02/07/2007) - beta
 ----------
 + Table action_config has been added.
 + Config actions (color and alias) has been added to the fron-end.
 + Remote host can be controlled. (use syslogd -r option)
 + New Filters to check windows login has been added. (use Remote Host with ntsyslog)
 * Fix install process when create /var/www/auth2db Dir.

ver. 0.2.1 (23/06/2007) - beta
 ----------
 + Control panel has been added to the front-end (config: hosts, users, alerts, smtp).
 + A new install and configure proccess has been added (exec auth2db-config).
 + Alerts checks has been added (to activate, change Y/n in auth2db.conf).
 + New tables has been added for Alerts, Hosts config, Smtp config and Users config.
 + Login has been added to the front-end.
 + Alerts by email has been added. (configure smtp in front-end control panel)

ver. 0.2.0 (12/06/2007) - beta
 ----------
 + A graphics report has been added to the front-end (php gd is needed). 
 + Advanced and Simple views have been added to the front-end. 
 + There are new options to configure the filters in the front-end. 
 + An auth2db -v option for program details has been added. 
 * There is a fix for filters.conf. 
 * There is a fix for updateip functions (sshd and smbd sessions).

ver. 0.1.9 (06/06/2007) - alpha
 ----------
 + Regex searches has been added. 
 + Filter regexes can be configured in filters.conf. 
 + A user can create her own regex search for any service. 
 + Individual log configuration can be done in filters.conf for all types. 
 - Log configuration from auth2db.conf has been removed. 
 + Enabled and Disabled types were added in filters.conf. 
 - Code has been cleaned.

ver. 0.1.8 (01/06/2007) - alpha
 ----------
 * Flag register insert after MySQL insert has been fixed. 
 + A check error validator before MySQL inserts has been added. 
 + An error log has been added. 
 + Email configuration in auth2db-config has been added. 
 + Sending email to the admin if an error occurred has been added. 
 + A graphic design has been added to the front-end.

ver. 0.1.7 (27/05/2007) - alpha
 ----------
 + Type PROFTPD control has been added.
 * FIX for APACHE control date parser.
 + HIDE columns option ("machine" and "detail"), has been added to the PHP Front-end.
 + Statistic has been updated at the PHP front-end.

ver. 0.1.6 (26/05/2007) - alpha
 ----------
 + Type APACHE control has been added.
 * FIX auth2db-config recognize the linux distribution and suggest the auth log path.
 + Statistic at the PHP front-end has been added.

ver. 0.1.5 (23/05/2007) - alpha
 ----------
 * MySQLdb connection has been validated.
 * auth2db-config recognize the linux distribution and suggest the auth log path.

ver. 0.1.4 (22/05/2007) - alpha
----------
 + CONFIG_AUTH_FILE to change the path of "auth.log", has been added 
    (in /etc/auth2db/auth2db.conf)
 + hostname with socket.gethostnmae() in auth2db, has been added 
    (general method to know the hostname in differents distributions)

ver. 0.1.3 (21/05/2007) - alpha
----------
 + Daemon has been added
 * Fix PHP front-end flags images with file_exists()

ver. 0.1.2 (20/05/2007) - alpha
----------
 + Daemon has been added
 + A new installer in python has been added. (python setup.py install) 
 + A session_start, at the PHP front-end, has been added.

ver. 0.1.1 (18/05/2007) - alpha
----------
 * Se crea el flag...dat si no existe

www/encode.js
function urlDecode(str){
    str=str.replace(new RegExp('\\+','g'),' ');
    return unescape(str);
}
function urlEncode(str){
    str=escape(str);
    str=str.replace(new RegExp('\\+','g'),'%2B');
    return str.replace(new RegExp('%20','g'),'+');
}
 
var END_OF_INPUT = -1;
 
var base64Chars = new Array(
    'A','B','C','D','E','F','G','H',
    'I','J','K','L','M','N','O','P',
    'Q','R','S','T','U','V','W','X',
    'Y','Z','a','b','c','d','e','f',
    'g','h','i','j','k','l','m','n',
    'o','p','q','r','s','t','u','v',
    'w','x','y','z','0','1','2','3',
    '4','5','6','7','8','9','+','/'
);
 
var reverseBase64Chars = new Array();
for (var i=0; i < base64Chars.length; i++){
    reverseBase64Chars[base64Chars[i]] = i;
}
 
var base64Str;
var base64Count;
function setBase64Str(str){
    base64Str = str;
    base64Count = 0;
}
function readBase64(){    
    if (!base64Str) return END_OF_INPUT;
    if (base64Count >= base64Str.length) return END_OF_INPUT;
    var c = base64Str.charCodeAt(base64Count) & 0xff;
    base64Count++;
    return c;
}
function encodeBase64(str){
    setBase64Str(str);
    var result = '';
    var inBuffer = new Array(3);
    var lineCount = 0;
    var done = false;
    while (!done && (inBuffer[0] = readBase64()) != END_OF_INPUT){
        inBuffer[1] = readBase64();
        inBuffer[2] = readBase64();
        result += (base64Chars[ inBuffer[0] >> 2 ]);
        if (inBuffer[1] != END_OF_INPUT){
            result += (base64Chars [(( inBuffer[0] << 4 ) & 0x30) | (inBuffer[1] >> 4) ]);
            if (inBuffer[2] != END_OF_INPUT){
                result += (base64Chars [((inBuffer[1] << 2) & 0x3c) | (inBuffer[2] >> 6) ]);
                result += (base64Chars [inBuffer[2] & 0x3F]);
            } else {
                result += (base64Chars [((inBuffer[1] << 2) & 0x3c)]);
                result += ('=');
                done = true;
            }
        } else {
            result += (base64Chars [(( inBuffer[0] << 4 ) & 0x30)]);
            result += ('=');
            result += ('=');
            done = true;
        }
        lineCount += 4;
        if (lineCount >= 76){
            result += ('\n');
            lineCount = 0;
        }
    }
    return result;
}
function readReverseBase64(){   
    if (!base64Str) return END_OF_INPUT;
    while (true){      
        if (base64Count >= base64Str.length) return END_OF_INPUT;
        var nextCharacter = base64Str.charAt(base64Count);
        base64Count++;
        if (reverseBase64Chars[nextCharacter]){
            return reverseBase64Chars[nextCharacter];
        }
        if (nextCharacter == 'A') return 0;
    }
    return END_OF_INPUT;
}
 
function ntos(n){
    n=n.toString(16);
    if (n.length == 1) n="0"+n;
    n="%"+n;
    return unescape(n);
}
 
function decodeBase64(str){
    setBase64Str(str);
    var result = "";
    var inBuffer = new Array(4);
    var done = false;
    while (!done && (inBuffer[0] = readReverseBase64()) != END_OF_INPUT
        && (inBuffer[1] = readReverseBase64()) != END_OF_INPUT){
        inBuffer[2] = readReverseBase64();
        inBuffer[3] = readReverseBase64();
        result += ntos((((inBuffer[0] << 2) & 0xff)| inBuffer[1] >> 4));
        if (inBuffer[2] != END_OF_INPUT){
            result +=  ntos((((inBuffer[1] << 4) & 0xff)| inBuffer[2] >> 2));
            if (inBuffer[3] != END_OF_INPUT){
                result +=  ntos((((inBuffer[2] << 6)  & 0xff) | inBuffer[3]));
            } else {
                done = true;
            }
        } else {
            done = true;
        }
    }
    return result;
}
 
var digitArray = new Array('0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f');
function toHex(n){
    var result = ''
    var start = true;
    for (var i=32; i>0;){
        i-=4;
        var digit = (n>>i) & 0xf;
        if (!start || digit != 0){
            start = false;
            result += digitArray[digit];
        }
    }
    return (result==''?'0':result);
}
 
function pad(str, len, pad){
    var result = str;
    for (var i=str.length; i<len; i++){
        result = pad + result;
    }
    return result;
}
 
function encodeHex(str){
    var result = "";
    for (var i=0; i<str.length; i++){
        result += pad(toHex(str.charCodeAt(i)&0xff),2,'0');
    }
    return result;
}
 
var hexv = {
  "00":0,"01":1,"02":2,"03":3,"04":4,"05":5,"06":6,"07":7,"08":8,"09":9,"0A":10,"0B":11,"0C":12,"0D":13,"0E":14,"0F":15,
  "10":16,"11":17,"12":18,"13":19,"14":20,"15":21,"16":22,"17":23,"18":24,"19":25,"1A":26,"1B":27,"1C":28,"1D":29,"1E":30,"1F":31,
  "20":32,"21":33,"22":34,"23":35,"24":36,"25":37,"26":38,"27":39,"28":40,"29":41,"2A":42,"2B":43,"2C":44,"2D":45,"2E":46,"2F":47,
  "30":48,"31":49,"32":50,"33":51,"34":52,"35":53,"36":54,"37":55,"38":56,"39":57,"3A":58,"3B":59,"3C":60,"3D":61,"3E":62,"3F":63,
  "40":64,"41":65,"42":66,"43":67,"44":68,"45":69,"46":70,"47":71,"48":72,"49":73,"4A":74,"4B":75,"4C":76,"4D":77,"4E":78,"4F":79,
  "50":80,"51":81,"52":82,"53":83,"54":84,"55":85,"56":86,"57":87,"58":88,"59":89,"5A":90,"5B":91,"5C":92,"5D":93,"5E":94,"5F":95,
  "60":96,"61":97,"62":98,"63":99,"64":100,"65":101,"66":102,"67":103,"68":104,"69":105,"6A":106,"6B":107,"6C":108,"6D":109,"6E":110,"6F":111,
  "70":112,"71":113,"72":114,"73":115,"74":116,"75":117,"76":118,"77":119,"78":120,"79":121,"7A":122,"7B":123,"7C":124,"7D":125,"7E":126,"7F":127,
  "80":128,"81":129,"82":130,"83":131,"84":132,"85":133,"86":134,"87":135,"88":136,"89":137,"8A":138,"8B":139,"8C":140,"8D":141,"8E":142,"8F":143,
  "90":144,"91":145,"92":146,"93":147,"94":148,"95":149,"96":150,"97":151,"98":152,"99":153,"9A":154,"9B":155,"9C":156,"9D":157,"9E":158,"9F":159,
  "A0":160,"A1":161,"A2":162,"A3":163,"A4":164,"A5":165,"A6":166,"A7":167,"A8":168,"A9":169,"AA":170,"AB":171,"AC":172,"AD":173,"AE":174,"AF":175,
  "B0":176,"B1":177,"B2":178,"B3":179,"B4":180,"B5":181,"B6":182,"B7":183,"B8":184,"B9":185,"BA":186,"BB":187,"BC":188,"BD":189,"BE":190,"BF":191,
  "C0":192,"C1":193,"C2":194,"C3":195,"C4":196,"C5":197,"C6":198,"C7":199,"C8":200,"C9":201,"CA":202,"CB":203,"CC":204,"CD":205,"CE":206,"CF":207,
  "D0":208,"D1":209,"D2":210,"D3":211,"D4":212,"D5":213,"D6":214,"D7":215,"D8":216,"D9":217,"DA":218,"DB":219,"DC":220,"DD":221,"DE":222,"DF":223,
  "E0":224,"E1":225,"E2":226,"E3":227,"E4":228,"E5":229,"E6":230,"E7":231,"E8":232,"E9":233,"EA":234,"EB":235,"EC":236,"ED":237,"EE":238,"EF":239,
  "F0":240,"F1":241,"F2":242,"F3":243,"F4":244,"F5":245,"F6":246,"F7":247,"F8":248,"F9":249,"FA":250,"FB":251,"FC":252,"FD":253,"FE":254,"FF":255
};
 
function decodeHex(str){
    str = str.toUpperCase().replace(new RegExp("s/[^0-9A-Z]//g"));
    var result = "";
    var nextchar = "";
    for (var i=0; i<str.length; i++){
        nextchar += str.charAt(i);
        if (nextchar.length == 2){
            result += ntos(hexv[nextchar]);
            nextchar = "";
        }
    }
    return result;
    
}